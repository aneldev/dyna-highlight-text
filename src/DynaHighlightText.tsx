import * as React from "react";

export interface IDynaHighlightTextProps {
  sourceText: string;
  highlightWords: string;
  caseSensitive?: boolean;
  onHighlightRender?: (word: string) => TPartialText;
}

export type TPartialText = JSX.Element | string;

export const DynaHighlightText = (props: IDynaHighlightTextProps): JSX.Element => {
  const {
    sourceText,
    highlightWords,
    caseSensitive = false,
    onHighlightRender = word => <u>{word}</u>,
  } = props;

  const highlightText = (): TPartialText[] => {
    const textOriginal: string = sourceText;
    let textProcess: string = sourceText;
    const wordsOriginal: string[] = highlightWords.split(' ').filter(w => !!w);
    let wordsProcess: string[] = wordsOriginal.concat();
    const output: TPartialText[] = [];
    let partialOutput: string = '';

    if (!caseSensitive) {
      textProcess = textProcess.toLowerCase();
      wordsProcess = wordsProcess.map(word => word.toLowerCase());
    }

    for (let iChar: number = 0; iChar < textProcess.length; iChar++) {
      let wordFound: boolean = false;
      for (let iWord: number = 0; iWord < wordsProcess.length; iWord++) {
        let word: string = wordsProcess[iWord];
        if (textProcess.substring(iChar, iChar + word.length) === word) {
          wordFound = true;
          if (partialOutput) {
            output.push(partialOutput);
            partialOutput = '';
          }
          let hWord: TPartialText = onHighlightRender(textOriginal.substring(iChar, iChar + word.length));
          if (typeof hWord !== 'string') hWord = React.cloneElement(hWord, {key: iChar + '-' + iWord});
          output.push(hWord);
          iChar += wordsOriginal[iWord].length - 1;
          break;
        }
      }
      if (!wordFound) partialOutput += textOriginal[iChar];
    }
    if (partialOutput) output.push(partialOutput);
    return output;
  }

  return <>{highlightText()}</>;
}
