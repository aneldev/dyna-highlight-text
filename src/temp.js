function highlightText(text, words, wholeWords, caseSensitive, cbRenderHighlighedWord) {
  let textOriginal = text;
  let textProcess = text;
  let wordsOriginal = words.split(' ').filter(w => !!w);
  let wordsProcess = [].concat(wordsOriginal);
  let output = '';

  if (!caseSensitive) {
    textProcess = textProcess.toLowerCase();
    wordsProcess = wordsProcess.map(word => word.toLowerCase());
  }

  for (var iChar = 0; iChar < textProcess.length; iChar++) {
    let wordFound = false;
    for (var iWord = 0; iWord < wordsProcess.length; iWord++) {
      let word = wordsProcess[iWord];
      if (textProcess.substr(iChar, word.length) === word) {
        wordFound = true;
        output += cbRenderHighlighedWord(wordsOriginal[iWord]);
        iChar += wordsOriginal[iWord].length - 1;
        break;
      }
    }
    if (!wordFound) output += textOriginal[iChar];
  }
}
