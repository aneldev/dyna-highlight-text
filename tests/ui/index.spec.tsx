import * as React from "react";
import * as enzyme from "enzyme";

// @ts-ignore
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

enzyme.configure({ adapter: new Adapter() });

import {DynaHighlightText} from '../../src';

describe('Home', () => {
  let wrapper;

  it('should highlight the word at front', () => {
    wrapper = enzyme.mount(
      <DynaHighlightText sourceText="Secret world" highlightWords="secret"/>,
      {}
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should highlight the word at the end', () => {
    wrapper = enzyme.mount(
      <DynaHighlightText sourceText="Secret world" highlightWords="world"/>,
      {}
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should highlight the word at the end', () => {
    wrapper = enzyme.mount(
      <DynaHighlightText sourceText="The secret world" highlightWords="Secret"/>,
      {}
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should highlight two words', () => {
    wrapper = enzyme.mount(
      <DynaHighlightText sourceText="The secret world" highlightWords="world Secret"/>,
      {}
    );
    expect(wrapper).toMatchSnapshot();
  });

});
