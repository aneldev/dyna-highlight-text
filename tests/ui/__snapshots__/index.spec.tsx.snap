// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should highlight the word at front 1`] = `
<DynaHighlightText
  highlightWords="secret"
  sourceText="Secret world"
>
  <u
    key="0-0"
  >
    Secret
  </u>
   world
</DynaHighlightText>
`;

exports[`Home should highlight the word at the end 1`] = `
<DynaHighlightText
  highlightWords="world"
  sourceText="Secret world"
>
  Secret 
  <u
    key="7-0"
  >
    world
  </u>
</DynaHighlightText>
`;

exports[`Home should highlight the word at the end 2`] = `
<DynaHighlightText
  highlightWords="Secret"
  sourceText="The secret world"
>
  The 
  <u
    key="4-0"
  >
    secret
  </u>
   world
</DynaHighlightText>
`;

exports[`Home should highlight two words 1`] = `
<DynaHighlightText
  highlightWords="world Secret"
  sourceText="The secret world"
>
  The 
  <u
    key="4-1"
  >
    secret
  </u>
   
  <u
    key="11-0"
  >
    world
  </u>
</DynaHighlightText>
`;
